Fixes #4551 - GCE provisioning support

dLobatog opened the pull request:
Enable provisioning of VMs through Google Compute Engine. Volume-wise,
this is currently limited to creating a VM with an attached disk that
contains the image specified. Future enhancements should include
choosing any available disks to auto-attach the VM and not force the
user to create a new one through Foreman.

I thought about removing disks when removing the VM as well, since disks are created at the same time as the VM. However users might attach their own disks through the GCE console and we don't want to destroy these.

dLobatog committed:
Fixes #4551 - GCE provisioning support

Enable provisioning of VMs through Google Compute Engine. Volume-wise,
this is currently limited to creating a VM with an attached disk that
contains the image specified. Future enhancements should include
choosing any available disks to auto-attach the VM and not force the
user to create a new one through Foreman.

dLobatog commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -100,6 +116,14 @@ def self.provider_friendly_name
       "Google"
     end
 
+    def interfaces_attrs_name
```
This is the interfaces attr_name used by Fog, however we can't really use it at this point.

dLobatog commented:
Latest push fixes 2 rubocop warnings

unorthodoxgeek commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -78,6 +85,15 @@ def available_images
+        :name         => "foreman-#{Time.now.to_i}",
+        :size_gb      => size.to_i,
+        :zone_name    => zone,
+        :source_image => client.images.select { |i| i.id == image_uuid }.first.name
```
is there a way to find an image by uuid without fetching all images? in other apis there is a find method (or they allow finders inside the all method)

dLobatog replied:
Not by ID, only by name, which doesn't really work in our case because the image model stores the UUID.

unorthodoxgeek commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -100,6 +116,14 @@ def self.provider_friendly_name
     end
 
+    def interfaces_attrs_name
+      "network_interfaces"
```
very tiny nitpick - use a symbol. this one will return a new string instance every time you call this method.

unorthodoxgeek commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -119,7 +143,7 @@ def check_google_key_path
     def vm_instance_defaults
       super.merge(
-        :zone_name => zone,
+        :zone      => zone,
```
can we stop using these horrible indented items? every change causes horrible merge issues just because it "looks better on the eye" (it doesn't, IMHO)
should be

``` ruby
:zone => zone,
:long_name_for_example => "whatever"
```

ares replied:
+1, this is not puppet (luckily :-)

unorthodoxgeek commented on app/models/concerns/fog_extensions/google/server.rb:
```diff
@@ -12,12 +12,20 @@ def flavors
 
       def image_id
-        image_name unless disks.nil?
+        image_name unless disks.empty?
```
use #blank?
empty is (by default) defined in Enumerable.
nil.empty? is defined in ActiveSupport 3.2, but it might get deprecated.

dLobatog commented:
@unorthodoxgeek Thanks, updated. Change the labels if you can please :smile:  (or ask for access) - normally the PR sender shouldn't change its own PRs

unorthodoxgeek commented:
I'm okay with the code. looking good. not really sure how to test it manually, don't have access to GCE resources...

dLobatog commented:
[test]

domcleal commented on app/views/compute_resources_vms/form/gce/_volume.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div class="fields">
+  <div class="form-group">
+    <%= text_f f, :size_gb, :class => "col-md-2", :label => _("Size (GB)"), :onchange => 'capacity_edit(this)', :help_inline => 'A boot disk of this size containing the Operating System image will be created' %>
```
missing extraction

dLobatog commented:
@domcleal Thanks, string extracted.

unorthodoxgeek commented:
@domcleal - I think this is ready for merge.

domcleal commented:
Has anybody tested it?  I don't see any comments.

mmoll commented:
Somebody on IRC reported this to work. Maybe this can even find its way into 1.8.1?

domcleal commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -78,6 +85,15 @@ def available_images
 
+    def create_disk(size, image_uuid)
+      client.disks.create(
+        :name         => "foreman-#{Time.now.to_i}",
```
shouldn't this use the VM name?  We use Time.now in vm_instance_defaults, but it's overwritten somewhere else with the real VM name later on in the process - this doesn't look like it is.

domcleal commented on app/models/concerns/fog_extensions/google/server.rb:
```diff
@@ -12,12 +12,20 @@ def flavors
         pretty_machine_type
       end
+
+      def stop
```
no need for these, 1.29.0 supports them: https://github.com/fog/fog/commit/2114827fb7fd977734e29f3569a21965e1cfe231

domcleal commented on app/views/compute_resources_vms/form/gce/_volume.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div class="fields">
+  <div class="form-group">
+    <%= text_f f, :size_gb, :class => "col-md-2", :label => _("Size (GB)"), :onchange => 'capacity_edit(this)', :help_inline => _('A boot disk of this size containing the Operating System image will be created') %>
```
Nitpick, lower case "Operating System"

domcleal commented:
I hit the issue [mentioned here](https://github.com/theforeman/foreman/pull/2017#issuecomment-67931172) about Fog::VERSION being required when adding a new GCE CR, adding `require 'fog/version'` into fog_extensions below fog/google fixed it for me.  (This won't be an issue in Fog 1.30/2 as GCE support's extracted and it was fixed.)

On the new host form, I have two storage sections, is that right?  I had to put a size in either the bottom one (or both) to make it create the OS disk with a non-zero size.

![screenshot from 2015-05-06 09 09 23](https://cloud.githubusercontent.com/assets/482089/7489565/14dc8106-f3d0-11e4-91af-b76a81f6c4b8.png)

domcleal commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -39,10 +39,8 @@ def email=(email)
 
-    #TODO: allow to select public / private ip address that foreman tries to find
-    def provided_attributes
-      super.merge({ :ip => :public_ip_address })
```
Are you sure you should remove this?  When it's removed, Foreman makes the _user_ supply the IP address, which won't be appropriate as it has to come from one of their configured subnets.  With it here, Foreman pulls the IP address from the GCE VM.

domcleal replied:
also still relevant I think

domcleal commented:
> I hit the issue mentioned here about Fog::VERSION being required when adding a new GCE CR, adding require 'fog/version' into fog_extensions below fog/google fixed it for me. (This won't be an issue in Fog 1.30/2 as GCE support's extracted and it was fixed.)

We're now on Fog 1.30.0, this should have fixed itself.

dLobatog commented:
Latest update addresses all comments in the review except for the two Storage sections, I'm looking into that and testing why would it create an empty disk instead of a disk of the size of the OS.

dLobatog commented:
[test]

dLobatog commented:
Removed the double 'storage' section. It seems like I either add a network section or someone without a proxy would have to input fake settings in interfaces? I don't know why a subnet or domain would be required for this

domcleal commented:
I'm not sure if a network section is necessary.  Setting a domain is needed for the FQDN of the host, but setting an IP and subnet should be optional, if the provided_attributes is put back to supply the IP from the VM to the host.

domcleal commented:
@eLobato could you take a look at this?  https://github.com/eLobato/foreman/compare/4551...domcleal:4551-gce

Feel free to squash it in.  What I've done is handle all disks submitted from the form, so it can now actually create more than just the first boot disk, and initialises the form with a reasonable disk size.  Also fixed the networking requirement for IP and it should pull back the correct one (although I can't seem to create a VM with an external IP).

I can't get the auto-deletion of disks to work - haven't checked yet to see if the API call I'm making to set the auto-delete flag is persisting or not.

dLobatog commented:
Squashed, thanks!

domcleal commented on app/models/concerns/fog_extensions/google/server.rb:
```diff
@@ -12,12 +12,22 @@ def flavors
+      end
+
+      def vm_ip_address
+        external_ip ? public_ip_addres : private_ip_address
```
typo here, public_ip_..."address"

domcleal commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -60,20 +63,53 @@ def zone=(zone)
-      super(args.merge(ssh))
+      vm = super(args.merge(ssh))
+
+      args[:disks].each { |disk| vm.set_disk_auto_delete(true, disk.name) }
```
did you get this to work?  It still seems to leave orphaned disks here.

domcleal replied:
Setting back to WoC due to this comment.

dLobatog commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -60,20 +63,52 @@ def zone=(zone)
       ssh      = { :username => username, :public_key => key_pair.public }
-      super(args.merge(ssh))
+      vm = super(args.merge(ssh))
+      vm.disks.each { |disk| vm.set_disk_auto_delete(true, disk['deviceName']) }
```
I've updated this line, it was using `args[:disks]` names, but it turns out `vm.disks` uses a slightly different naming. This prevents leaving orphaned disks after an instance is deleted.

domcleal replied:
Nice fix!

domcleal commented:
Merged as 1fa8dcf, thanks @eLobato.

I'm seeing an error using the external IP address tickbox, but I'll file that as a new issue as it's different to volume support.
